### PR TITLE
feat(cli): scaffold out `astro add` command

### PR DESCRIPTION
**Pull request by natemoo-re**
## Changes

- Introduces a new `astro add` command to automatically configure `integrations`.
- `astro add preact --yes` will skip any prompts, automatically make changes. Thanks @JuanM04!
- `astro add --help` will print help instructions. Thanks @JuanM04!

- See demo below:
  - `npx astro add tailwind` is already installed, so wizard stops early
  - `npx astro add preact` is not installed, configuration is updated and dependencies installed
  - `npx astro add` has no arguments, so an interactive wizard prompts users to select frameworks then other integrations

https://user-images.githubusercontent.com/7118177/160188381-926d0624-c781-4f09-ba2c-242a526b828b.mov

## Testing

Manually tested, interactive CLI

## Docs

[Integrations Guide updated](https://github.com/withastro/docs/pull/258)

### CODE REVIEW

File: packages/astro/src/cli/add.ts
Comment (FredKSchott) [inline packages/astro/src/cli/add.ts:62]: nice! I was actually going to suggest that it would make sense to limit our launch support to only config files with a single `export default defineConfig()`. Looks like you've already done that here!

File: packages/astro/src/transform/babel.ts
Comment (FredKSchott) [inline packages/astro/src/transform/babel.ts:20]: how bad is the code change without this? Mind sharing a screenshot? 

The reason is that for every user that appreciates this formatting, there are probably more who don't have a prettier config that this will mess with in a way that will be hard to revert (since it includes the needed code change).

Comment (FredKSchott) [reply packages/astro/src/transform/babel.ts:20]: also fwiw I just learned that recast preserves original formatting. If we still have rough alignment to migrate to recast long-term, then I'd feel better about this as the short-term fix.

Comment (natemoo-re) [reply packages/astro/src/transform/babel.ts:20]: It's definitely bad without it, so I'm not sure which option is worse.

Screen Shot 2022-03-24 at 4 54 37 PM

Comment (FredKSchott) [reply packages/astro/src/transform/babel.ts:20]: those imports! My eyes!

File: packages/astro/src/transform/babel.ts
Comment (FredKSchott) [inline packages/astro/src/transform/babel.ts:11]: safe to remove to kill this dep entirely?

File: packages/astro/src/transform/babel.ts
Comment (natemoo-re) [reply packages/astro/src/transform/babel.ts:20]: Actually, just kidding. It's not as bad as I thought. Comments get a little wonky but there's not much we can do about that.

Screen Shot 2022-03-24 at 4 59 35 PM
.

Comment (FredKSchott) [reply packages/astro/src/transform/babel.ts:20]: oh! that's really not bad at all. strong +1 for that over prettier.

File: packages/astro/src/cli/add.ts
Comment (FredKSchott) [inline packages/astro/src/cli/add.ts:265]: I'd add a note saying something like "If you skip this step, you can always run it yourself later" or so that people who know they need to run something else don't feel like you're exiting the program.

File: packages/astro/src/cli/index.ts
Comment (FredKSchott) [inline packages/astro/src/cli/index.ts:116]: I don't think this project root behavior was ever documented, and I've never heard of anyone using it. Either add a note here to remove before v1.0, or go ahead and remove it now.

**Comment by JuanM04** [comment]
@natemoo-re I love the logging you've made! I've added some spinners, what do you think?

File: packages/astro/src/cli/index.ts
Comment (FredKSchott) [inline packages/astro/src/cli/index.ts:144]: why not run this after loading config? I'd imagine that could also act as a light validation/lint step of your config before we make changes to it.

Also, should we use `projectRoot` or `process.cwd()` always when we run things like `npm install`? I wonder if I'd actually expect it to use `process.cwd()` as a user?

File: packages/astro/src/core/config.ts
Comment (FredKSchott) [inline packages/astro/src/core/config.ts:270]: refactor loadConfig() to call this internally? seems like a lot of copy-paste would be removed

File: packages/astro/src/transform/babel.ts
Comment (FredKSchott) [inline packages/astro/src/transform/babel.ts:1]: nit on `src/transform` directory: is this meant to be generally usable across the rest of the codebase? I like the idea of keeping this new code a bit separate/quarantined from the rest of the codebase until it has a chance to settle, given its last-minute nature.

Thoughts on `src/core/add` to match `src/core/dev|preview|build` ?

File: packages/astro/src/transform/imports.ts
Comment (FredKSchott) [inline packages/astro/src/transform/imports.ts:31]: just checking if this is still the case, feel free to resolve if so

File: packages/astro/src/cli/add.ts
Comment (FredKSchott) [inline packages/astro/src/cli/add.ts:61]: ah wait, I guess we are handling the no `defineExport` case here, and `ensureImport` is adding the import if its missing.

Thoughts on removing this and just erroring if `defineConfig` isn't found, with a message saying `"migrate to defineConfig to enable the add command"`?

Otherwise, When this code path does run, even if it works I imagine it being a bit of a jarring diff for your entire config object to become red/green in the diff preview (also lots more formatting risk with such a big number of lines changed). 

Bonus: this would remove a lot of Babel code that you then wouldn't need to migrate to recast later :)

File: packages/astro/src/transform/imports.ts
Comment (natemoo-re) [reply packages/astro/src/transform/imports.ts:31]: Yes, unfortunately `insertAfter` has a tendency to separate comments from the nearest node and create noisy diffs.

File: packages/astro/src/cli/add.ts
Comment (natemoo-re) [reply packages/astro/src/cli/add.ts:61]: Ah I see what you're saying! That works for me.

File: packages/astro/src/cli/add.ts
Comment (FredKSchott) [inline packages/astro/src/cli/add.ts:74]: +1 for linking to the docs here. Something like:

```
Astro could not update your astro.config.js file safely.
Reason: ${err.message}

You will need to add these integration(s) manually.
Documentation: https://next--astro-docs-2.netlify.app/en/guides/integrations-guide/
```

File: packages/astro/src/core/config.ts
Comment (natemoo-re) [reply packages/astro/src/core/config.ts:270]: Yes I'd prefer this! This was supposed to just **resolve** the config but I had to rollback that change temporarily. Definitely want to refactor this, though.

File: packages/astro/src/cli/add.ts
Comment (FredKSchott) [reply packages/astro/src/cli/add.ts:74]: fwiw I think `https://next--astro-docs-2.netlify.app` is fine to link to while in `next` mode, if we can remember to update before official launch.

File: packages/astro/src/cli/add.ts
Comment (FredKSchott) [inline packages/astro/src/cli/add.ts:86]: Safe to move this into the same try-catch as above? As a user I think the same error message I just commented on above would make sense for both stages.

File: packages/astro/src/cli/add.ts
Comment (FredKSchott) [inline packages/astro/src/cli/add.ts:76]: `return` needed here? Also, here and elsewhere `return;` will most likely give you an error code of 0, when you really want 1. 

Possible to throw the original error with something like this?
```
catch (err) {
  err.message = `Astro could not update your astro.config.js file safely.
Reason: ${err.message}

You will need to add these integration(s) manually.
Documentation: https://next--astro-docs-2.netlify.app/en/guides/integrations-guide/`
throw err;
}
```

That at least lets you save the correct stack trace and get the exit code 1, but I'm not sure how much of an anti-pattern that is. Also okay to create a new error object here to throw.

File: packages/astro/src/cli/add.ts
Comment (FredKSchott) [inline packages/astro/src/cli/add.ts:93]: same as the error state above? As a user I'd be curious why this happened, what are my next steps, etc. 

To use the error message format from above comments, `Reason: cancelled.` would be enough there.

File: packages/astro/src/cli/add.ts
Comment (FredKSchott) [inline packages/astro/src/cli/add.ts:97]: super nit: `pnpm` `npm` and `yarn` all seem happy with the shorter "Configuration up-to-date." style message, without a "No changes needed!" at the end. Maybe okay to cut that?

File: packages/astro/src/cli/add.ts
Comment (FredKSchott) [inline packages/astro/src/cli/add.ts:132]: do we still support the "astro" config in your package.json through proload?

also, I'm assuming `astro add` doesn't support `astro.config.ts` since it would fail these parse steps. I think that's a fine limitation for now as long as we handle it gracefully.

File: packages/astro/src/cli/add.ts
Comment (natemoo-re) [reply packages/astro/src/cli/add.ts:93]: This only happens if you answer `n (no)` to the prompt, so it should be pretty obvious in context.

Comment (natemoo-re) [reply packages/astro/src/cli/add.ts:93]: This feels like a different state than `error` since the user is opting-out of continuing, so a red error state might be a little jarring.

File: packages/astro/src/cli/add.ts
Comment (FredKSchott) [inline packages/astro/src/cli/add.ts:286]: I know its another dep for this feature, but I'm a strong fan of `execa` over `exec` for reliability, which I think is at a premium here. In this case I'd rather start with the confidence choice and then optimize later.

For example, I was going to suggest piping the output here as a stream, so that it doesn't buffer and then throw like this. Also, combining stdout and sterr together, since I think logs can happen in either stream at any time, in both success and failure. Also, not sure if `cwd` is handled here.

Not sure how to do this with exec, but with execa its:

```js
import { execaCommand } from 'execa';

await execaCommand(cmd, { cwd, stdout: 'inherit', stderr: 'inherit' });
```

File: packages/astro/src/cli/add.ts
Comment (FredKSchott) [reply packages/astro/src/cli/add.ts:93]: fair enough! I left a few comments related to error reporting but feel free to resolve them as you see fit

File: packages/astro/src/cli/index.ts
Comment (JuanM04) [reply packages/astro/src/cli/index.ts:116]: I would put it under a `--project-root` flag

File: packages/astro/src/core/add/index.ts
Comment (JuanM04) [inline packages/astro/src/core/add/index.ts:52]: We could move this to a JSON file somewhere, like `packages/integrations/manifest.json`. This file could be used by `create-astro` too

File: packages/astro/src/core/add/index.ts
Comment (JuanM04) [inline packages/astro/src/core/add/index.ts:304]: Why do you return an object instead of a string?

**Review by JuanM04 — COMMENTED** [review]
The interactive prompt is great! I've added some comments

File: packages/astro/src/core/add/index.ts
Comment (natemoo-re) [reply packages/astro/src/core/add/index.ts:311]: Just to improve highlighting the message! Not super important, but a nice touch.

File: packages/astro/src/core/add/index.ts
Comment (natemoo-re) [reply packages/astro/src/core/add/index.ts]: A great idea! I'm not sure if a separate package is in scope for this PR, but it's a good follow-up step.

**Comment by FredKSchott** [comment]
A few of us just got a demo of this and I can confirm it's awesome. Thank you again @JuanM04 for kicking this effort off, and @natemoo-re for helping to carry it over the line.

I chatted with some of the other core maintainers, and the plan is to release this as "experimental" in the docs and any READMEs that reference the command. We'll use the next couple of weeks to gather feedback from users before formalizing support.

File: packages/astro/src/cli/add.ts
Comment (natemoo-re) [reply packages/astro/src/cli/add.ts]: I'm adding a specific state for `package.json#astro`. And `astro.config.ts` does work! Babel can handle it.

File: packages/astro/src/cli/add.ts
Comment (natemoo-re) [reply packages/astro/src/cli/add.ts]: I'm going to skip this for now, but noted! The current behavior is to add `defineConfig` automatically. The diff has been improved significantly to be much less noisy.

File: packages/astro/src/cli/index.ts
Comment (natemoo-re) [reply packages/astro/src/cli/index.ts]: Yep, removed.

File: packages/astro/src/cli/add.ts
Comment (natemoo-re) [reply packages/astro/src/cli/add.ts]: Updated this to all be handled in one place!

File: packages/astro/src/cli/add.ts
Comment (FredKSchott) [reply packages/astro/src/cli/add.ts]: Nice! I feel like we might want to drop the `package.json` support but I'm happy to push that convo to another time

File: packages/astro/src/cli/add.ts
Comment (FredKSchott) [reply packages/astro/src/cli/add.ts]: Sounds good! If it's already written, no harm in keeping it. If/when we rewrite to recast I'll bring it back up as a suggestion.

File: packages/astro/src/cli/index.ts
Comment (FredKSchott) [inline packages/astro/src/cli/index.ts:102]: Following up to a different thread: `projectRoot` here is either the `--project-root` value or undefined if no flag given. This is different from `astroConfig.projectRoot`. 

You mentioned that moving this to run after `loadConfig` with the other commands would require too big of a refactor. Can you add a comment here explaining why this command runs so early, for future refactor-ers? 

My main hesitation here is that calling this `projectRoot` makes it sound like the official config value, but its operating very different here. I actually think there's a bug in your code if you use `--project-root`, since flags.projectRoot should be a string but you have typed the `add()` argument as a URL. I could be wrong here, but I don't think so.

My suggestion would be to make this change to make it clear that this is a non-standard idea of cwd, and not an officially resolve projectRoot.

```suggestion
            const cwd = projectRoot ? path.resolve(projectRoot) : process.cwd();
			await add(packages, { cwd, flags, logging });
```

File: packages/astro/src/core/config.ts
Comment (natemoo-re) [reply packages/astro/src/core/config.ts]: Decided _not_ to tackle this now, but I agree that this is an obvious candidate for a refactor in the future.

File: packages/astro/src/cli/index.ts
Comment (natemoo-re) [reply packages/astro/src/cli/index.ts:102]: Thank you, great callout. Will update!

File: packages/astro/src/core/add/index.ts
Comment (FredKSchott) [inline packages/astro/src/core/add/index.ts:172]: - mentioned in person, but don't overwrite existing config files.
- also came up in person, can you log when this happens so that users aren't surprised?

File: packages/astro/src/core/add/index.ts
Comment (FredKSchott) [inline packages/astro/src/core/add/index.ts:173]: - mentioned in person, but don't overwrite existing config files.
- also came up in person, can you log when this happens so that users aren't surprised?

File: packages/astro/src/core/add/index.ts
Comment (FredKSchott) [inline packages/astro/src/core/add/index.ts:176]: - nit: missing a period at the end
- nit: can you log these dependencies by name? Something like:

```
Successfully added the following integrations to your project:
  - @astrojs/preact
  - @astrojs/tailwind
  - @astrojs/solid-js
```

File: packages/astro/src/core/add/index.ts
Comment (FredKSchott) [inline packages/astro/src/core/add/index.ts:207]: nit: if a function is expected to throw, `process.exit()`, etc. I really like to make make it explicit in the function name for future readers. Instead of `bail()`, what about `logPrettyErrorAndThrow()`?

Or, instead of throwing inside this function and doing the odd `return throw err` behavior in the callers, what about:

```diff
  // example usage below
  debug('add', 'Error parsing/modifying astro config: ', err);
- return bail(err as Error);
+ throw prettyCommandError(err);
```

less side-effects ftw!

File: packages/astro/src/core/add/index.ts
Comment (FredKSchott) [inline packages/astro/src/core/add/index.ts:306]: nit: `...` vs. `:`? also `these changes` vs. `the following changes`?

Slight preference for `Astro will make the following changes to your config file:`

Comment (FredKSchott) [reply packages/astro/src/core/add/index.ts:306]: This matches the `Astro will run the following command to install...` message below in format as well, which is a happy coincidence

**Review by FredKSchott — APPROVED** [review]
Made some comments but overall LGTM once those are addressed as needed

File: packages/astro/src/core/add/index.ts
Comment (natemoo-re) [reply packages/astro/src/core/add/index.ts]: Yep, this is much better. Updated!

File: packages/astro/src/cli/index.ts
Comment (natemoo-re) [reply packages/astro/src/cli/index.ts]: Updated to use `cwd` and pass through to use the behavior defined in the `core/config.ts` file.

**Review by bholmesdev — APPROVED** [review]
An invaluable `add`ition to the CLI 👏

**Pull request merged**
